cpp/src/arrow/engine/substrait/expression_internal.cc

```diff
@@ -57,45 +57,70 @@ Result<compute::Expression> FromProto(const substrait::Expression& expr) {
     case substrait::Expression::kSelection: {
       if (!expr.selection().has_direct_reference()) break;
 
-      util::optional<compute::Expression> root_expr;
+      util::optional<compute::Expression> out;
       if (expr.selection().has_expression()) {
-        ARROW_ASSIGN_OR_RAISE(root_expr, FromProto(expr.selection().expression()));
+        ARROW_ASSIGN_OR_RAISE(out, FromProto(expr.selection().expression()));
       }
 
-      const auto& ref = expr.selection().direct_reference();
-      switch (ref.reference_type_case()) {
-        case substrait::Expression::ReferenceSegment::kStructField: {
-          if (ref.struct_field().has_child()) break;
-
-          FieldRef out(ref.struct_field().field());
+      const auto* ref = &expr.selection().direct_reference();
+      while (ref != NULLPTR) {
```

Comment:
NULLPTR only needs to be used in public headers
```suggestion
      while (ref != nullptr) {
```

Reply:
56e4f1b